allennlp/tests/common/test_util.py

```diff
@@ -88,3 +103,30 @@ def test_get_frozen_and_tunable_parameter_names(self):
         ) = util.get_frozen_and_tunable_parameter_names(model)
         assert set(frozen_parameter_names) == {"linear.weight", "linear.bias"}
         assert set(tunable_parameter_names) == {"conv.weight", "conv.bias"}
+
+    def test_plugins_are_discovered_and_imported(self):
+        plugins_root = self.FIXTURES_ROOT / "plugins"
+        project_a_fixtures_root = plugins_root / "project_a"
+        project_b_fixtures_root = plugins_root / "project_b"
+
+        with tempfile.TemporaryDirectory() as temp_dir_a, tempfile.TemporaryDirectory() as temp_dir_b:
+            distutils.dir_util.copy_tree(project_a_fixtures_root, temp_dir_a)
```

Comment:
So, the way this works, you are required to name your project `allennlp_plugins`?  That seems suboptimal to me.  All of our example repos, and our recommended third-party setup, says "name your package whatever you want, then use `--include-package=[my_package_name]`".  Now you're saying that your package name *has* to be `allennlp_plugins`.  Right?

I agree that this is pretty easy on our side, but I don't really like the user side of it much.

Reply:
No, you're not required to name your project `allennlp_plugins`. Just having that folder.

Reply:
But only things in that folder will be registered.  And basically all of the code that you write in a typical simple project is stuff that you want to be registered.  So, basically, yes, your project.

Reply:
You can import your own project from that folder.

Reply:
I'll add it to the test to make it clearer.

Reply:
Check the new commit, the test shows how it works.

Essentially, the user has to:

    1. Add an `allennlp_plugins` folder.
    2. Add a folder inside of it, with the name of his project preferably.
    3. Add an `__init__.py` file that imports what he/she wants AllenNLP to see. He can add more files and Python symbols if she/he wants.

Maybe the `tree` output can make it clearer:

```
$ tree allennlp/tests/fixtures/plugins/project_a/
allennlp/tests/fixtures/plugins/project_a/
├── a
│   ├── a.py
│   └── __init__.py
└── allennlp_plugins
    └── a
        └── __init__.py

3 directories, 3 files
```

Reply:
Still seems a lot simpler to just not require the user to make that separate directory, and instead just add one line to a `.allennlp_plugins` file in the same directory (we could have our code check both the current directory, for local plugins, and a global `~/.allennlp/plugins` file for globally-installed stuff).  Or, if it's easy to make our pip packages add things to this namespace package, use that for first-party libraries, and use `.allennlp_plugins` for user libraries.

Reply:
I'd tend to choose what Python recommends for this, but what you propose sounds a bit simpler for the user for his local project, yeah. For the global scope, I'm not convinced of having that global file. Maybe we can support both if it works fine for first-party pip-installed libraries. Let me do more testing and get back to you.

Reply:
There are two problems here: (1) having things like `pip install allennlp-server` just work with our normal commands when you install them.  And (2) letting users remove the need for constant `--include-package` statements everywhere.  These problems seem very related to me, but it's possible that they both should have separate solutions, because of python environment issues.  That's ok.